client/src/send_and_confirm_transactions_in_parallel.rs

```diff
@@ -0,0 +1,476 @@
+        signers::Signers,
+        transaction::{Transaction, TransactionError},
+    },
+    solana_tpu_client::{
+        nonblocking::tpu_client::set_message_for_confirmed_transactions,
+        tpu_client::{Result, TpuSenderError},
+    },
+    std::{
+        sync::{
+            atomic::{AtomicU32, AtomicU64, Ordering},
+            Arc,
+        },
+        time::Duration,
+    },
+    tokio::{sync::RwLock, task::JoinHandle, time::Instant},
+};
+
+const BLOCKHASH_REFRESH_RATE: Duration = Duration::from_secs(10);
+const TPU_RESEND_REFRESH_RATE: Duration = Duration::from_secs(2);
+type QuicTpuClient = TpuClient<QuicPool, QuicConnectionManager, QuicConfig>;
+
+#[derive(Clone, Debug)]
+struct TransactionData {
+    last_valid_blockheight: u64,
+    message: Message,
+    index: usize,
+    serialized_transaction: Vec<u8>,
+}
+
+#[derive(Clone, Debug, Copy)]
+struct BlockHashData {
+    pub blockhash: Hash,
+    pub last_valid_blockheight: u64,
+}
+
+#[derive(Clone, Debug, Copy)]
+pub struct SendAndConfrimConfig {
+    pub with_spinner: bool,
+    pub resign_txs_count: Option<usize>,
+}
+
+pub fn send_and_confirm_transactions_in_parallel_blocking<T: Signers + ?Sized>(
+    rpc_client: Arc<BlockingRpcClient>,
+    tpu_client: Option<QuicTpuClient>,
+    messages: &[Message],
+    signers: &T,
+    config: SendAndConfrimConfig,
+) -> Result<Vec<Option<TransactionError>>> {
+    let fut = send_and_confirm_transactions_in_parallel(
+        rpc_client.get_inner_client().clone(),
+        tpu_client,
+        messages,
+        signers,
+        config,
+    );
+    tokio::task::block_in_place(|| rpc_client.runtime().block_on(fut))
+}
+
+fn create_blockhash_data_updating_task(
+    rpc_client: Arc<RpcClient>,
+    blockhash_data_rw: Arc<RwLock<BlockHashData>>,
+    current_blockheight: Arc<AtomicU64>,
+) -> JoinHandle<()> {
+    tokio::spawn(async move {
+        loop {
+            if let Ok((blockhash, last_valid_blockheight)) = rpc_client
+                .get_latest_blockhash_with_commitment(rpc_client.commitment())
+                .await
+            {
+                *blockhash_data_rw.write().await = BlockHashData {
+                    blockhash,
+                    last_valid_blockheight,
+                };
+            }
+
+            if let Ok(blockheight) = rpc_client.get_block_height().await {
+                current_blockheight.store(blockheight, Ordering::Relaxed);
+            }
+            tokio::time::sleep(BLOCKHASH_REFRESH_RATE).await;
+        }
+    })
+}
+
+fn create_transaction_confirmation_task(
+    rpc_client: Arc<RpcClient>,
+    current_blockheight: Arc<AtomicU64>,
+    unconfirmed_transasction_map: Arc<DashMap<Signature, TransactionData>>,
+    errors_map: Arc<DashMap<usize, TransactionError>>,
+    num_confirmed_transactions: Arc<AtomicU32>,
+) -> JoinHandle<()> {
+    tokio::spawn(async move {
+        // we also check last time all transaction that have just expired between two checks
```

Comment:
```suggestion
        // check transactions that are not expired or have just expired between two checks
```

This part is not clear,